mlflow/store/artifact/s3_artifact_repo.py

```diff
@@ -150,44 +179,135 @@ def _upload_file(self, s3_client, local_file, bucket, key):
-            if root != local_dir:
-                rel_path = os.path.relpath(root, local_dir)
-                rel_path = relative_path_to_artifact_path(rel_path)
-                upload_path = posixpath.join(dest_path, rel_path)
-            for f in filenames:
-                self._upload_file(
-                    s3_client=s3_client,
-                    local_file=os.path.join(root, f),
-                    bucket=bucket,
-                    key=posixpath.join(upload_path, f),
+    def _get_write_credential_infos(self, remote_file_paths):
+        """
+        Instead of returning ArtifactCredentialInfo objects, we instead return a list of initialized
+        S3 client. We do so because S3 clients cannot be instantiated within each thread.
+        """
+        return [self._get_s3_client() for _ in remote_file_paths]
+
+    def _upload_to_cloud(self, cloud_credential_info, src_file_path, artifact_file_path):
+        dest_path = posixpath.join(self.bucket_path, artifact_file_path)
+        key = os.path.normpath(dest_path)
+        if (
+            MLFLOW_ENABLE_MULTIPART_UPLOAD.get()
+            and os.path.getsize(src_file_path) > _MULTIPART_UPLOAD_CHUNK_SIZE
+        ):
+            self._multipart_upload(cloud_credential_info, src_file_path, self.bucket, key)
+        else:
+            self._upload_file(cloud_credential_info, src_file_path, self.bucket, key)
+
+    def _multipart_upload(self, cloud_credential_info, local_file, bucket, key):
+        # Create multipart upload
+        s3_client = cloud_credential_info
+        response = s3_client.create_multipart_upload(Bucket=bucket, Key=key)
+        upload_id = response["UploadId"]
+
+        # Create presigned URL for each part
+        num_parts = math.ceil(os.path.getsize(local_file) / _MULTIPART_UPLOAD_CHUNK_SIZE)
+
+        def _get_presigned_upload_part_url(part_number):
+            return s3_client.generate_presigned_url(
+                "upload_part",
+                Params={
+                    "Bucket": bucket,
+                    "Key": key,
+                    "UploadId": upload_id,
+                    "PartNumber": part_number,
+                },
+            )
+
+        presigned_urls = [
+            _get_presigned_upload_part_url(part_number)
+            for part_number in range(1, num_parts + 1)
+        ]
+
+        # define helper functions for uploading data
+        def _upload_part(presigned_url, data):
+            with cloud_storage_http_request("put", presigned_url, data=data) as response:
+                augmented_raise_for_status(response)
+                return response.headers["ETag"]
+
+        def _upload_part_retry(presigned_url, part_number, local_file, start_byte, size):
+            data = read_chunk(local_file, size, start_byte)
+            try:
+                return _upload_part(presigned_url, data)
+            except requests.HTTPError as e:
+                if e.response.status_code not in (401, 403):
```

Comment:
If ``s3:ListObjects`` permission isn't granted for this user, would this infinitely loop? Should there be a max retry break condition within this handler?

Reply:
Or will the ``None`` return value from an invalid presigned_url throw an appropriate Exception that won't be captured by the [401, 403] capture logic?

Reply:
It shouldn't. This function only retries twice. Though, I am wondering whether we should have a retry at all. Given that the presigned URLs are generated from the same S3 client initialized with a scoped token. So if one presigned URL's credential expires, I'm not sure that the next one would work either.

Reply:
@smurching actually I'm curious what you think about this. The presigned URLs created by the S3 client should work only as long as the temporary credentials used to create the client work. So if one presigned URL has a 401 error, it would not make sense to retry. Maybe it would make sense to retry for a 403 error.

Reply:
Yeah I don't think we need to retry using the same credentials for any 4xx errors